File: server/graph-store.ts
import { randomUUID } from 'node:crypto';
import { mkdir, readFile, readdir, rename, writeFile } from 'node:fs/promises';
import { resolve, join } from 'node:path';
import type { AgentRun, RunRequest, TaskGraph } from '../shared/types.js';
import { compileReadGraph, contractHash, digest, graphApplicable, graphSchema, orderedNodes, runReadGraph, taskKey } from './graph.js';
import { createRun } from './runtime.js';
import type { Tool } from './tools.js';
import { retrieveWorkflows } from './gagent.js';

export class GraphStore {
  private graphs = new Map<string, TaskGraph>();
  private queue: Promise<unknown> = Promise.resolve();
  constructor(private directory = resolve('artifacts/graphs')) {}
  async restore() {
    await mkdir(this.directory, { recursive: true, mode: 0o700 });
    for (const file of await readdir(this.directory)) {
      if (!/^[a-f\d-]{36}\.json$/.test(file)) continue;
      try {
        const graph = graphSchema.parse(JSON.parse(await readFile(join(this.directory, file), 'utf8'))) as TaskGraph;
        if (graph.id + '.json' === file) this.graphs.set(graph.id, graph);
      } catch { /* Invalid artifacts are never promoted into executable graphs. */ }
    }
  }
  list(scenario?: RunRequest['scenario'], source?: RunRequest['source']) {
    return [...this.graphs.values()].filter(graph => (!scenario || graph.scenario === scenario) && (!source || graph.source === source)).sort((a, b) => b.version - a.version || b.createdAt.localeCompare(a.createdAt));
  }
  get(id: string) { return this.graphs.get(id); }
  select(request: RunRequest, tools: Tool[]) {
    const candidates = this.list(request.scenario, request.source);
    const environment = environmentHash(request.source);
    for (const graph of candidates) if (!graphApplicable(graph, request, tools, environment)) return { graph, candidates: [], reason: undefined };
    return { graph: undefined, candidates: retrieveWorkflows(candidates, request, tools, environment), reason: candidates.length ? graphApplicable(candidates[0], request, tools, environment) || undefined : '尚无经过校验的历史读取图' };
  }
  async learn(run: AgentRun, tools: Tool[], signal: AbortSignal = AbortSignal.timeout(60000)): Promise<TaskGraph> {
    // Serialize learning so concurrent completions cannot assign duplicate versions.
    const operation = this.queue.then(async () => {
      const existing = this.list().find(graph => graph.sourceRunId === run.id && graph.contractHash === contractHash(tools) && graph.environmentHash === environmentHash(run.request.source));
      if (existing) return existing;
      signal.throwIfAborted();
      const begin = Date.now();
      const nodes = compileReadGraph(run, tools);
      orderedNodes(nodes, tools);
      const compileMs = Date.now() - begin;
      const probe = createRun({ ...run.request, strategy: 'react' }, null);
      probe.initial = structuredClone(run.initial); probe.state = structuredClone(run.initial);
      const context = { run: probe, signal, evidence: new Set<string>() };
      let calls = 0;
      const validationStart = Date.now();
      await runReadGraph(nodes, tools, {
        invoke: async (name, args) => {
          signal.throwIfAborted();
          if (++calls > 60) throw new Error('影子验证超过 60 次读取，拒绝发布任务图');
          const tool = tools.find(tool => tool.name === name)!;
          return tool.execute(args, context);
        }, node: () => {},
      }, signal);
      if (digest(probe.state) !== digest(probe.initial)) throw new Error('读取工具修改了业务状态，拒绝学习');
      const key = taskKey(run.request.task);
      const preceding = this.list(run.request.scenario, run.request.source).filter(graph => graph.taskKey === key);
      const graph: TaskGraph = {
        id: randomUUID(), version: Math.max(0, ...preceding.map(graph => graph.version)) + 1,
        createdAt: new Date().toISOString(), sourceRunId: run.id, scenario: run.request.scenario, source: run.request.source,
        task: run.request.task, taskKey: key, contractHash: contractHash(tools), environmentHash: environmentHash(run.request.source), nodes,
        sourceModelRequests: run.metrics.modelRequests, sourceInputTokens: run.metrics.inputTokens, sourceOutputTokens: run.metrics.outputTokens,
        validation: { status: 'passed', toolCalls: calls, durationMs: Date.now() - validationStart, compileMs, learningModelRequests: 0 },
        scope: 'read-prefix', compilerVersion: 1,
      };
      graphSchema.parse(graph);
      await mkdir(this.directory, { recursive: true, mode: 0o700 });
      const path = join(this.directory, graph.id + '.json');
      await writeFile(path + '.tmp', JSON.stringify(graph, null, 2), { mode: 0o600 });
      await rename(path + '.tmp', path);
      this.graphs.set(graph.id, graph);
      return graph;
    });
    this.queue = operation.catch(() => {});
    return operation;
  }
}

export function environmentHash(source: RunRequest['source']) {
  // Instance URL participates in the guard; secrets never enter the graph file.
  const url = source === 'sandbox' ? 'synthetic-v1' : process.env[source.toUpperCase() + '_BASE_URL'] || 'unconfigured';
  return digest({ source, url });
}
export const graphStore = new GraphStore();

File: backend/graph_store.py
import asyncio
from copy import deepcopy
import json
import os
from pathlib import Path
import time
from uuid import uuid4, UUID
from . import config
from .domain import now
from .autotool import digest
from .graph import compile_read_graph, ordered_nodes, run_read_graph, contract_hash, task_key, applicable
from .gagent import retrieve
from .tools import ToolContext
from .runtime import create_run


def environment_hash(source):
    return digest({'source': source, 'url': 'synthetic-v1' if source == 'sandbox' else os.getenv(source.upper() + '_BASE_URL', 'unconfigured'), 'runtime': 'python-v1'})


def write_private(path, value):
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    temporary = path.with_name(path.name + '.' + str(uuid4()) + '.tmp')
    descriptor = os.open(temporary, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    with os.fdopen(descriptor, 'w') as stream:
        stream.write(json.dumps(value, ensure_ascii=False, indent=2) + '\n' if not isinstance(value, str) else value)
    os.replace(temporary, path)


class GraphStore:
    def __init__(self, directory=None):
        self.directory = Path(directory or config.ARTIFACTS / 'graphs-python')
        self.graphs = {}
        self.lock = asyncio.Lock()

    def restore(self):
        self.directory.mkdir(parents=True, exist_ok=True, mode=0o700)
        for path in self.directory.glob('*.json'):
            try:
                graph = json.loads(path.read_text())
                UUID(graph['id'])
                if path.stem == graph['id'] and graph.get('scope') == 'read-prefix' and graph.get('validation', {}).get('status') == 'passed' and isinstance(graph.get('nodes'), list):
                    self.graphs[graph['id']] = graph
            except (ValueError, KeyError, TypeError):
                continue

    def list(self, scenario=None, source=None):
        return sorted([g for g in self.graphs.values() if (not scenario or g['scenario'] == scenario) and (not source or g['source'] == source)], key=lambda g: (g['version'], g['createdAt']), reverse=True)

    def select(self, request, tools):
        candidates = self.list(request['scenario'], request['source'])
        environment = environment_hash(request['source'])
        for graph in candidates:
            if not applicable(graph, request, tools, environment):
                ordered_nodes(graph['nodes'], tools)
                return {'graph': graph, 'candidates': [], 'reason': None}
        return {'graph': None, 'candidates': retrieve(candidates, request, tools, environment), 'reason': '尚无精确适用的已验证读取图'}

    async def learn(self, run, tools):
        async with self.lock:
            existing = next((g for g in self.list() if g['sourceRunId'] == run['id'] and g['contractHash'] == contract_hash(tools) and g['environmentHash'] == environment_hash(run['request']['source'])), None)
            if existing:
                return existing
            started = time.monotonic()
            nodes = compile_read_graph(run, tools)
            ordered_nodes(nodes, tools)
            compile_ms = round((time.monotonic() - started) * 1000)
            probe = create_run(dict(run['request'], strategy='react'))
            probe['initial'], probe['state'] = deepcopy(run['initial']), deepcopy(run['initial'])
            context, calls = ToolContext(probe), 0
            by_name = {t.name: t for t in tools}
            async def invoke(name, args, node):
                nonlocal calls
                calls += 1
                if calls > 60:
                    raise ValueError('影子验证超过 60 次读取')
                return await by_name[name].execute(args, context)
            validation_start = time.monotonic()
            await asyncio.wait_for(run_read_graph(nodes, tools, invoke), timeout=60)
            if probe['initial'] != probe['state']:
                raise ValueError('读取工具修改了业务状态，拒绝学习')
            key = task_key(run['request']['task'])
            version = max([0] + [g['version'] for g in self.list(run['request']['scenario'], run['request']['source']) if g['taskKey'] == key]) + 1
            graph = {'id': str(uuid4()), 'version': version, 'createdAt': now(), 'sourceRunId': run['id'], 'scenario': run['request']['scenario'], 'source': run['request']['source'],
                     'task': run['request']['task'], 'taskKey': key, 'contractHash': contract_hash(tools), 'environmentHash': environment_hash(run['request']['source']), 'nodes': nodes,
                     'sourceModelRequests': run['metrics']['modelRequests'], 'sourceInputTokens': run['metrics']['inputTokens'], 'sourceOutputTokens': run['metrics']['outputTokens'],
                     'validation': {'status': 'passed', 'toolCalls': calls, 'durationMs': round((time.monotonic() - validation_start) * 1000), 'compileMs': compile_ms, 'learningModelRequests': 0},
                     'scope': 'read-prefix', 'compilerVersion': 1, 'backend': 'python'}
            write_private(self.directory / (graph['id'] + '.json'), graph)
            self.graphs[graph['id']] = graph
            return graph
